jq_properties_parser.go

```diff
@@ -0,0 +1,104 @@
+package opslevel_jq_parser
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/rs/zerolog/log"
+
+	"github.com/opslevel/opslevel-go/v2024"
+)
+
+type JQPropertiesParser struct {
+	assigns []*JQFieldParser
+}
+
+func NewJQPropertiesParser(cfg PropertyRegistrationConfig) *JQPropertiesParser {
+	assigns := make([]*JQFieldParser, len(cfg.Assign))
+	for i, expression := range cfg.Assign {
+		assigns[i] = NewJQFieldParser(expression)
+	}
+	return &JQPropertiesParser{
+		assigns: assigns,
+	}
+}
+
+func propertyAssignmentFromMap(definition string, input map[string]interface{}) (opslevel.PropertyInput, error) {
+	var owner string
+	var value opslevel.JsonString
+	log.Debug().Msgf("property assignment '%s' input is %#v", definition, input)
+	for k, v := range input {
+		switch k {
+		case "owner":
+			owner = v.(string)
+		case "value":
+			jsonString, err := opslevel.NewJSONInput(v)
+			if err != nil {
+				return opslevel.PropertyInput{}, err
+			}
+			value = *jsonString
+		}
+	}
+	if owner != "" && value != "" {
+		return opslevel.PropertyInput{
+			Owner:      *opslevel.NewIdentifier(owner),
+			Definition: *opslevel.NewIdentifier(definition),
+			Value:      value,
+		}, nil
+	}
+	log.Warn().Msgf("got incomplete property assignment '%s'", definition)
+	return opslevel.PropertyInput{}, nil
+}
+
+func (p *JQPropertiesParser) parse(programs []*JQFieldParser, data string) ([]opslevel.PropertyInput, error) {
+	output := make([]opslevel.PropertyInput, 0, len(programs))
+	for _, program := range programs {
+		response, err := program.Run(data)
+		if err != nil {
+			log.Warn().Msg("properties parser: jq error")
+			continue
+		}
+		if response == "" {
+			log.Warn().Msg("properties parser: jq returned empty string")
+			continue
+		}
+
+		if strings.HasPrefix(response, "[") && strings.HasSuffix(response, "]") {
+			var properties []map[string]string
+			if err := json.Unmarshal([]byte(response), &properties); err != nil {
+				log.Warn().Msg("skipping a property assignment - error decoding at start")
+				continue
+			}
+			for _, item := range properties {
+				if len(item) != 1 {
+					log.Warn().Msg("skipping a property assignment - bad format")
+					continue
+				}
+				var def string
+				for k := range item {
+					def = k
+				}
+				var propertyBody map[string]interface{}
+				if err := json.Unmarshal([]byte(item[def]), &propertyBody); err != nil {
+					log.Warn().Msgf("got error decoding property assignment '%s'", def)
+					continue
+				}
+				out, err := propertyAssignmentFromMap(def, propertyBody)
+				if err != nil {
+					log.Warn().Msgf("got error parsing property assignment '%s'", def)
+					continue
+				}
+				output = append(output, out)
+			}
+		}
```

Comment:
breaking this logic down into re-useable function so you can test the inputs and outputs of that function will be easier to grok.  Especially whenever you see a for loop inside a for loop if you make the logic inside the forloop a function call you can have 1 error handler after the function all handle logging all the messages and running continue and then inside the function call you can just use early `return` statements.  

All of this together will make reviewing the logic flow WAY easier.

Reply:
The logic in this entire file is way shorter, added early `continue` statements. There is only one nested `for` which is necessary to extract a single key (prop def alias) from a map. (I could also use `maps.Keys()` if we really want to remove that `for`)